### PR TITLE
Examples fixed

### PR DESCRIPTION
**Pull request by gelvidge**
**Summary**
Fixes majority of html examples.  Some small bugs may remain but 99% ok

**Description for the changelog**
Updates example html files and fixes bugs in core code that were found when testing

**Other info**

covers #73

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -46,4 +47,5 @@
     "webpack-merge": "^5.8.0"
   },
   "sideEffects": true
+
```
Comment (tbouffard) [inline package.json:50]: Extra change

File: packages/core/src/serialization/Codec.ts [hunk 2]
```diff
@@ -339,6 +342,20 @@ class Codec {
    * @param into Optional object to be decodec into.
    */
   decode(node: Element, into?: any): any {
+
+
+    //****Edit note - there may be other classes needed here????******
```
Comment (tbouffard) [inline packages/core/src/serialization/Codec.ts:347]: ??

File: packages/core/src/serialization/Codec.ts [hunk 3]
```diff
@@ -347,11 +364,11 @@ class Codec {
 
       try {
         // @ts-ignore
-        ctor = window[node.nodeName];
+        ctor = classMap[node.nodeName]
+        if (ctor == undefined) console.log(node.nodeName)
```
Comment (tbouffard) [inline packages/core/src/serialization/Codec.ts:368]: Please do proper Error management

File: packages/core/src/types.ts [hunk 4]
```diff
@@ -82,6 +82,7 @@ export type CellStateStyle = {
   imageBorder?: ColorValue;
   imageHeight?: number;
   imageWidth?: number;
+  imageVerticalAlign?: VAlignValue;
```
Comment (tbouffard) [inline packages/core/src/types.ts:85]: For indicator only? There was no need for this in the original mxGraph code, why do we need this now?

File: packages/core/src/util/cloneUtils.ts [hunk 5]
```diff
@@ -25,11 +25,11 @@ export const clone = function _clone(obj: any, transients: string[] | null=null,
         i != ObjectIdentity.FIELD_NAME &&
         (transients == null || transients.indexOf(i) < 0)
       ) {
-        if (!shallow && typeof obj[i] === 'object') {
+       if (!shallow && typeof obj[i] === 'object') {
```
Comment (tbouffard) [inline packages/core/src/util/cloneUtils.ts:28]: Extra change + formatting issue

File: packages/core/src/util/styleUtils.ts [hunk 6]
```diff
@@ -438,9 +440,10 @@ export const setCellStyles = (
 
         if (cell) {
           const style = cell.getStyle();
-          style[key] = value;
+          const styleClone =clone(style)
```
Comment (tbouffard) [inline packages/core/src/util/styleUtils.ts:443]: Why changing this?

File: packages/core/src/view/Graph.ts [hunk 7]
```diff
@@ -580,10 +580,8 @@ class Graph extends EventSource {
       const newParent = change.child.getParent();
       this.view.invalidate(change.child, true, true);
 
-      if (
-        newParent &&
-        (!this.getDataModel().contains(newParent) || newParent.isCollapsed())
-      ) {
+      if (newParent == null || !this.getDataModel().contains(newParent) || newParent.isCollapsed())
```
Comment (tbouffard) [inline packages/core/src/view/Graph.ts:583]: Let's review mxgraph original code or explain why you need to change the logic here.

File: packages/core/src/view/canvas/SvgCanvas2D.ts [hunk 8]
```diff
@@ -994,9 +994,10 @@ class SvgCanvas2D extends AbstractCanvas2D {
    * Private helper function to create SVG elements
    */
   roundrect(x: number, y: number, w: number, h: number, dx: number, dy: number) {
-    if (!this.node) return;
```
Comment (tbouffard) [inline packages/core/src/view/canvas/SvgCanvas2D.ts:997]: Check original mxgraph implementation, I don't understand why we need to change the logic

File: packages/core/src/view/handler/EdgeHandler.ts [hunk 9]
```diff
@@ -1579,10 +1579,13 @@ class EdgeHandler {
 
                 edge = cloned;
               }
-
               edge = this.connect(edge, terminal, this.isSource, clone, me);
+
```
Comment (tbouffard) [inline packages/core/src/view/handler/EdgeHandler.ts:1583]: extra break line

File: packages/core/src/view/handler/SelectionHandler.ts [hunk 10]
```diff
@@ -208,7 +208,7 @@ class SelectionHandler implements GraphPlugin {
    * Specifies if other cells should be used for snapping the right, center or
    * left side of the current selection. Default is false.
    */
-  guidesEnabled = false;
+  guidesEnabled = true;
```
Comment (tbouffard) [inline packages/core/src/view/handler/SelectionHandler.ts:211]: Why changing defaults? this impacts all users of the lib and not only examples

File: packages/core/src/view/GraphDataModel.ts [hunk 11]
```diff
@@ -389,7 +389,7 @@ export class GraphDataModel extends EventSource {
    * @param {Cell} cell  that represents the possible layer.
    */
   isLayer(cell: Cell) {
-    return this.isRoot(cell.getParent());
+    return cell != null  && this.isRoot(cell.getParent());
```
Comment (junsikshim) [inline packages/core/src/view/GraphDataModel.ts:392]: If `cell` can be null or undefined, I think it's better change the method signature,
perhaps something like `isLayer(cell?: Cell)`.

File: packages/core/src/view/handler/SelectionHandler.ts (hunk 10 above)
Comment (junsikshim) [inline packages/core/src/view/handler/SelectionHandler.ts:211]: If `guidesEnabled` should be `true` by default, please update the comments above. :)

File: packages/core/src/view/handler/SelectionHandler.ts [hunk 12]
```diff
@@ -300,7 +300,7 @@ class SelectionHandler implements GraphPlugin {
   /**
    * Maximum number of cells for which live preview should be used.  Default is 0 which means no live preview.
    */
-  maxLivePreview = 0;
+  maxLivePreview = 50;
```
Comment (junsikshim) [inline packages/core/src/view/handler/SelectionHandler.ts:303]: Like above, please change the comment accordingly if the default value must change.

File: packages/html/stories/Clipboard.stories.js [hunk 13]
```diff
@@ -175,54 +167,52 @@ const Template = ({ label, ...args }) => {
...
-
         // Merges existing layers and adds new layers
         graph.model.beginUpdate();
-        try {
-          for (let i = 0; i < childCount; i++) {
-            let parent = model.getRoot().getChildAt(i);
-
+							
+        
+          for (var i = 0; i < childCount; i++)
```
Comment (junsikshim) [inline packages/html/stories/Clipboard.stories.js:186]: Let's get rid of `var`s. :)

File: packages/html/stories/ExtendCanvas.stories.js [hunk 14]
```diff
@@ -28,8 +28,8 @@ const Template = ({ label, ...args }) => {
   const container = document.createElement('div');
   container.style.position = 'relative';
   container.style.overflow = 'auto';
-  container.style.width = `${args.width}px`;
-  container.style.height = `${args.height}px`;
+  container.style.width = '321px';
+  container.style.height = '241px';
```
Comment (junsikshim) [inline packages/html/stories/ExtendCanvas.stories.js:32]: It'd be great if we can set the pixel values using storybook arguments.

**Review by junsikshim — COMMENTED** [review]
Nice work!

I had to remove `NODE_OPTIONS=--openssl-legacy-provider` to make it run though, people with older versions of node (< 17) would face this problem.

A couple of things to comment on.
- Please run prettier on all the files so that the styles stay the same.
- I think IDE configuration files should be excluded from git repository(add to .gitignore), but it's up for debate.

Thanks!

File: packages/html/package.json [hunk 15]
```diff
@@ -4,7 +4,7 @@
   "description": "",
   "main": "index.js",
   "scripts": {
-    "dev": "start-storybook -p 8901"
+    "dev": "cross-env NODE_OPTIONS=--openssl-legacy-provider start-storybook -p 8901 ./public"
```
Comment (tbouffard) [inline packages/html/package.json:7]: Please revert this: see #74 and https://github.com/maxGraph/maxGraph/discussions/64#discussioncomment-1968814

File: packages/core/src/util/Utils.ts [hunk 16]
```diff
@@ -60,6 +60,22 @@ export const mixInto = (dest: any) => (mixin: any) => {
   }
 };
 
+//getBBox of SVG element before rendered in DOM
```
Comment (tbouffard) [inline packages/core/src/util/Utils.ts:63]: please write proper JSDoc

File: packages/core/src/view/canvas/SvgCanvas2D.ts [hunk 17]
```diff
@@ -1819,9 +1820,9 @@ class SvgCanvas2D extends AbstractCanvas2D {
         }
 
         bbox = new Rectangle(
-          (x + 1) * s.scale,
+          x  * s.scale,
           (y + 2) * s.scale,
-          w * s.scale,
+          (w +1) * s.scale,
           (h + 1) * s.scale
         );
```
Comment (tbouffard) [inline packages/core/src/view/canvas/SvgCanvas2D.ts:1827]: I am pretty sure the previous implementation cames from the orignal mxGraph.
There are a lot of place where there are `+1` on position, width or height. It is very hard to do the review here. Unfortunately, I don't have time to dig into the 56 commits to find which example may need the change and how this change impacts other examples.

File: packages/core/src/view/mixins/ConnectionsMixin.ts [hunk 18]
```diff
@@ -265,8 +265,7 @@ const ConnectionsMixin: PartialType = {
     let dy = 0;
 
     if (point) {
-      perimeter = edge.style[source ? 'exitPerimeter' : 'entryPerimeter'] || false;
-
+      perimeter = edge.style[source ? 'exitPerimeter' : 'entryPerimeter'] || true;
```
Comment (tbouffard) [inline packages/core/src/view/mixins/ConnectionsMixin.ts:268]: please don't change defaults

File: packages/core/src/view/mixins/FoldingMixin.ts [hunk 19]
```diff
@@ -92,7 +92,7 @@ type PartialFolding = Pick<
   | 'updateAlternateBounds'
 >;
 type PartialType = PartialGraph & PartialFolding;
-
+Client.imageBasePath = '/images'
```
Comment (tbouffard) [inline packages/core/src/view/mixins/FoldingMixin.ts:95]: You are changing the configuration here and this impacts all users of the lib, not only the example.

**Review by tbouffard — CHANGES_REQUESTED** [review]
Thanks for sharing this Pull Request. This is a lot of work.
I am pretty sure that the Pull Request contains a lot of valuable work but it is too large to review.
It also contains too much unwanted formatting changes that makes it hard to read.

There are also some logic changes that requires to be discussed because I don't understand why they are needed. I don't have time to look at all commits to try to find where they could be required.
The lib has no automatic tests, and examples don't cover all use cases. It is also very easy to fix an example and break another one. The current proposal also contains several default changes that need to be discussed.

I suggest to split the work into several smaller Pull Requests: start with examples that only require storybook files changes. That way, we can review and merge working examples quickly and have more time to discuss tricky subjects

**Comment by gelvidge** [comment]
Thanks both for reviewing and thanks for the feedback. I'll look into seeing how the PR can be improved by splitting and tidying up some of the formatting

**Comment by tbouffard** [comment]
@gelvidge hi, I will have some time on Thursday (June, 9th) to work on the project. I can try to cherry-pick some commits that quickly fix the examples in a dedicated PR.
Please let me know

**Comment by aminosbh** [comment]
@gelvidge we are currently on the process of changing the license of this project to the standard Apache 2.0 license (see #89). Do you agree to use the Apache 2.0 license on your contribution ?

**Comment by gelvidge** [comment]
Yes that is OK. I will aim to get around to submitting the new commits soon but all my contributions can be under Apache v2

**Comment by tbouffard** [comment]
@gelvidge hi, do you still plan to recreate some PR based on the fixes proposed here?
@mayorovad started fixing some stories and some are already merged. That's why there is some merge conflicts here.

So I suggest you synchronize together if you both want to set new stories.

**Comment by tbouffard** [comment]
This PR is kept opened on purpose even if it won't be merged. There is no need to update it, nor rebase it, as it it too large.

However, it contains a lot of valuable information and fixes. It can be used as a source of information to create new focused and small PR about a single topic.

**Comment by tbouffard** [comment]
Closing this PR.
It will be used to pick some changes when fixing remaining problems in stories. See #505.

**Pull request closed without merging**